docs/developer-docs/multi-chain/ethereum/evm-rpc/costs.mdx

```diff
@@ -0,0 +1,79 @@
+---
+keywords: [advanced, ethereum, evm rpc, evm, costs]
+---
+
+import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
+
+# EVM RPC costs
+
+<MarkdownChipRow labels={["Advanced", "Ethereum", "Tutorial"]} />
+
+## Overview
+
+Each call made to the EVM RPC canister costs [cycles](/docs/current/developer-docs/getting-started/cycles/cycles-faucet). These costs are paid by attaching cycles to each RPC call by using the `--with-cycles` flag when calling the EVM RPC canister with [dfx](/docs/current/developer-docs/developer-tools/cli-tools/cli-reference/dfx-canister#dfx-canister-call).
+
+## Costs
+
+JSON-RPC requests typically cost between 10^8 and 10^9 cycles, which is equivalent to approximately $0.0001 - $0.001 USD.
+
+The following formula shows how to calculate the cycles cost for an RPC request:
+
+```
+(
+  3M
+  + 60K * nodes_in_subnet
+  + 400 * request_size
+  + 800 * max_response
+  + provider_cost
+) * nodes_in_subnet
+```
+
+In order to accurately measure the HTTPS outcall cost and protect from an accidental large responses, the caller must specify the maximum expected number of bytes. Due to the potentially high cost of performing outcalls with a suboptimal max response size, it’s generally in the caller’s best interest to choose a value for `maxResponseBytes` on a case-by-case basis for each situation.
+
+Developers may choose to repeatedly call this method with an increasingly large value for `maxResponseBytes` (as performed by ckETH) to handle variable response sizes. We defer this choice of strategy to the caller. Various retry strategies could eventually be included as another canister method, although this is out of scope for the MVP. It’s worth noting that the caller essentially pays for each outcall as though they are sending these requests from their own canister.
+
+Additionally, the `RpcConfig` parameter supports an optional `responseSizeEstimate` which can be fine-tuned to reduce the cost of the RPC requests.
+
+Here is an approximate cost breakdown in USD for an RPC request on the Ethereum mainnet using Cloudflare Web3 and assuming a 13-node subnet:
+
+```
+JSON-RPC API request: $0.0001
+HTTPS outcalls: $0.001 (assuming 1kB request and 1kB response)
+Base cost: $0.0008
+JSON Request: $0.00008 / kB
+JSON Response: $0.00008 / kB
+```
+
+## Attaching the correct amount of cycles
+
+To determine how many cycles need to be sent with your RPC call, you can use the `requestCost` query method:
+
+```
+`requestCost`
+
+requestCost : (
+  source : JsonRpcSource,
+  jsonRequest : text,
+  maxResponseBytes :  nat64
+) -> (
+  Result<nat, RpcError>
+) query;
+```
+
+This query method accepts the same arguments as the canister's [`request`](evm-rpc.mdx#custom-json-rpc-requests) and returns the number of cycles to send with an equivalent call to the `request` method.
```

Comment:
> what is the strategy for calls to other methods the evm rpc canister offers, e.g. eth_get_logs?

Because the "Candid-RPC" methods use built-in retries and multiple providers, the amount of cycles required for each RPC call isn't predictable beforehand. The idea for these methods is to send a maximum cycles budget to perform the request.

> can we point developers to strategies that help them find the right maxResponseBytes.

We could mention that the developer can determine this value by measuring the size of expected JSON responses from an API playground such as [Alchemy Sandbox](https://sandbox.alchemy.com/). 

> why do we suddenly have to specify maxResponseBytes, while for the convenience methods (e.g. eth_get_logs) it's not necessary

This is because the convenience methods use a built-in default response bytes, which you can override with the `RpcConfig` value. 

> how do the convenience methods behave in contrast to request?

Essentially, the EVM RPC canister tries the request with increasing values of `maxResponseBytes` until it succeeds, starting with a reasonable default for each RPC method. In contrast, the JSON-RPC canister only ever makes one request, deferring the retry logic to the caller.

Reply:
> The idea for these methods is to send a maximum cycles budget to perform the request.

does that mean to send nearly all of the cycles available to the canister or is there an upper bound that we can communicate to developers?

Reply:
There is no built-in upper bound, so the canister will continue retrying until it runs out of cycles. This is important for requests such as `eth_getLogs`, where one request might return 100x as much data than the same request a few days earlier if there is a spike in smart contract activity. As a result, the optimal cycles budget is different for each use case. However, we could maybe suggest sending 10_000_000_000 cycles as a starting point and adjusting from there.

Reply:
Added Ryan's comment to the doc

Reply:
>  so the canister will continue retrying until it runs out of cycles.
with "the canister" you're referring to the evm rpc canister? 
and with "run out of cycles" you mean running out of cycles that have been attached to the message by the caller?

so in the worst case, i attach e.g. 10bn cycles, the response size is too big for the default value set by the evm rpc canister, so
- the canister keeps retrying with bigger response size (every calls cost being deducted from the cycles i sent with the message) until
a) the cycles attached to the message call have been used up -> call fails with out of cycles error, i can try again with more cycles and the last max response value as an argument (if its returned in the error message)
b) the max response size hits the message size limit -> call fails with message size exceeded error, i can try to limit the amount of blocks i want to scrape logs for if possible, otherwise there is nothing i can do